### DISCUSSION TITLE
Prevent GLSP elements from overlapping

### DISCUSSION THREAD

**Discussion by vatsal-uppal-1997** [eclipse-glsp/glsp#1014] [Q&A]
Hi All,

I was wondering if there is a way to prevent elements on the GLSP graphical editor (rectangular nodes and connecting lines) from overlapping?

Right now if I place a rectangular node on top of a rectangular node they both would overlap, is there a way I can prevent that from happening?

**Comment by tortmayr** [comment]
Hi @vatsal-uppal-1997,
The [`MovementRestrictor` API](https://github.com/eclipse-glsp/glsp-client/blob/fe0b29761e9844d8541ce4cafeaf7e00976157e1/packages/client/src/features/change-bounds/movement-restrictor.ts#L27) is an optional service that us used by tools to validate whether a certain move operation is allowed. 
In addition, we also provide a [`NoOverlapMovementRestrictor`](https://github.com/eclipse-glsp/glsp-client/blob/fe0b29761e9844d8541ce4cafeaf7e00976157e1/packages/client/src/features/change-bounds/movement-restrictor.ts#L47) implementation that should cover the behavior you described.
You can bind it in your diagram DI config with: 
```ts
bind(IMovementRestrictor).to(NoOverlapMovementRestrictor).inSingletonScope();
```
Note that there [is a known issue](https://github.com/eclipse-glsp/glsp/issues/394) with the `NoOverlapMovementRestrictor` and nested/structured nodes.

**Reply by vatsal-uppal-1997** [reply]
Hi @tortmayr,

Thanks for your reply! I tried what you suggested and it seems to do the job :D

Just two more follow up questions:

We have a drag and drop setup for GLSP, so can we detect collisions on drop somehow (so user cannot drop stuff on top of already existing elements)?

Can we prevent the edges from overlapping as well? Right now we're seeing something like this:

```
------
        |
          ------>
        |
------
```

Can we prevent the two edges from merging to a single point?
